### PR TITLE
Personalized points goal

### LINKED ISSUE #1332: Allow setting personal points goals for modules
Development idea from a student:

Käyttöliittymässä on palkki, joka näyttää, kuinka paljon pisteitä opiskelija on saanut. Jos on saanut kaikki pisteet, palkki on vihreä, mutta jos pisteet ovat vajaat, palkki on punainen tai oranssi.

Opiskelija voisi itse asettaa itselleen jonkin prosenttiosuuden, jota hän tavoittelee, ja kun se on täynnä, palkki olisi vihreä. Tällä hetkellä, kun vihreän palkin saa vain täysillä pisteillä, se ohjaa täydellisyyteen, vaikka monella kurssilla arvosanaan 5 riittää 90% pisteistä. Kaikki eivät myöskään tavoittele korkeinta arvosanaa, jolloin olisi hyvä, että tuota rajaa saisi muokattua.

image

### PR DESCRIPTION
**Pull request by mikaelGusse**
# Description

**What?**

Add a feature to the progress bar for modules which allows you to set a points goal for that particular module. When this goal is reached, the points progress bar changes color to blue.
image
image
image
image

**Why?**

Was asked for and improves the student experience of using APlus.

**How?**

Changed templates, added a new class to store points goals, added a modal to the interface to modify and create goals.
Also changed how the points required vertical bar is shown to be clearer.

Fixes #1332

# Testing

**Remember to add or update unit tests for new features and changes.**

* How to [test your changes in A-plus](https://github.com/apluslms/a-plus/tree/master/doc#running-tests-and-updating-translations)
* How to [test accessibility](https://wiki.aalto.fi/display/EDIT/How+to+check+the+accessibility+of+pull+requests)

**What type of test did you run?**

- [x] Accessibility test using the [WAVE](https://wave.webaim.org/extension/) extension.
- [x] Django unit tests.
- [x] Selenium tests.
- [x] Other test. *(Add a description below)*
- [x] Manual testing.

Added playwright tests to test this feature. Manually tested different combinations of required points, points goals etc. 

**Did you test the changes in**

- [ ] Chrome
- [x] Firefox
- [ ] This pull request cannot be tested in the browser.

**Think of what is affected by these changes and could become broken**

# Translation

- [x] Did you modify or add new strings in the user interface? ([Read about how to create translation](https://github.com/apluslms/a-plus/tree/master/doc#running-tests-and-updating-translations))

# Programming style

- [x] Did you follow our [style guides](https://apluslms.github.io/contribute/styleguides/)?
- [x] Did you use Python type hinting in all functions that you added or edited? ([type hints](https://docs.python.org/3/library/typing.html) for function parameters and return values)

# Have you updated the README or other relevant documentation?

- [ ] documents inside the doc directory.
- [ ] README.md.
- [ ] Aplus Manual.
- [ ] Other documentation (mention below which documentation).

# Is it Done?

- [ ] Reviewer has finished the code review
- [ ] After the review, the developer has made changes accordingly
- [ ] Customer/Teacher has accepted the implementation of the feature

*Clean up your git commit history before submitting the pull request!*

### CODE REVIEW

File: exercise/static/exercise/personalized_points_goal.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
+$(document).ready(function() {
+    const $pointsGoalForm = $('#pointsGoalForm');
+    document.getElementById('id_personalized_points_goal_input').focus();
+    $pointsGoalForm.on('submit', function(event) {
+        event.preventDefault();
+        $.ajax({
+            type: 'POST',
+            url: $pointsGoalForm.attr('action'),
+            data: $pointsGoalForm.serialize(),
+            success: function(response) {
```
Comment (ihalaij1) [inline exercise/static/exercise/personalized_points_goal.js:10]: Should we check if the value for the points goal is invalid, i.e., not a digit or a percentage, before sending it to the server? Then there would be less bad requests that have to be processed by the backend.

We would still double-check validity again on the server, and making sure that the value is above `points-to-pass` can be done in the backend only, if it is hard to do it in the frontend.

File: locale/fi/LC_MESSAGES/django.po [hunk 2]
```diff
@@ -4443,7 +4478,7 @@ msgstr "Aloita uudelleenarviointi"
 
 #: exercise/templates/exercise/staff/_resubmit_modal.html
 msgid "RE-SUBMIT_TO_SERVICE"
-msgstr "Lähetä arvosteltavaksi uudelleen"
+msgstr "Lähetä arvosteltavaksi uudelleen"#: exercise/templates/exercise/_user_results.html
```
Comment (ihalaij1) [inline locale/fi/LC_MESSAGES/django.po:4481]: Why is the text `#: exercise/templates/exercise/_user_results.html` after the message string?

File: locale/en/LC_MESSAGES/django.po [hunk 3]
```diff
@@ -1253,6 +1253,40 @@ msgstr "Late submissions are allowed until %(deadline)s. "
 msgid "LATE_SUBMISSION_POINTS_WORTH -- %(percent)s"
 msgstr "However, points are only worth %(percent)s%%. "
 
+#: exercise/templates/exercise/_user_results.html
+msgid "POINTS_GOAL"
+msgstr "Points goal"
+
+#: exercise/templates/exercise/_user_results.html
+msgid "POINTS_GOAL_TOOLTIP"
+msgstr "Set a personal points goal for this module. This goal is only visible to you. Once the goal has been reached, the points progress bar changes color to blue."
```
Comment (ihalaij1) [inline locale/en/LC_MESSAGES/django.po:1262]: Other translations spell "color" as "colour", could you change it?

File: locale/en/LC_MESSAGES/django.po [hunk 4]
```diff
@@ -1253,6 +1253,40 @@ msgstr "Late submissions are allowed until %(deadline)s. "
...
+
+#: exercise/templates/exercise/save_personalized_points_goal.html
+msgid "PERSONALIZED_POINTS_MODAL_TITLE"
+msgstr "Set personalized points goal for module"
+
+#: exercise/templates/exercise/save_personalized_points_goal.html
+msgid "LABEL_POINTS_GOAL_INPUT"
+msgstr "Input personalized goal as percentage or points (e.g., 50% or 150)"
```
Comment (ihalaij1) [inline locale/en/LC_MESSAGES/django.po:1270]: `as percentage` should be `as a percentage`.

File: locale/en/LC_MESSAGES/django.po [hunk 5]
```diff
@@ -1253,6 +1253,40 @@ msgstr "Late submissions are allowed until %(deadline)s. "
...
+
+#: exercise/templates/exercise/save_personalized_points_goal.html
+msgid "PERSONALIZED_POINTS_MODAL_FAILURE"
+msgstr "Failed to set personalized points goal"
+
+msgid "PERSONALIZED_POINTS_GOAL"
+msgstr "Points goal"
+
+msgid "PERSONALIZED_POINTS_MODAL_LESS_THAN_REQUIRED"
+msgstr "You cannot set the personalized points goal to be less than the points required to pass"
```
Comment (ihalaij1) [inline locale/en/LC_MESSAGES/django.po:1288]: Can you add `the module` at the end of the message?

File: locale/en/LC_MESSAGES/django.po [hunk 6]
```diff
@@ -1253,6 +1253,40 @@ msgstr "Late submissions are allowed until %(deadline)s. "
...
+msgid "PERSONALIZED_POINTS_MODAL_TITLE"
+msgstr "Set personalized points goal for module"
+
+#: exercise/templates/exercise/save_personalized_points_goal.html
+msgid "LABEL_POINTS_GOAL_INPUT"
+msgstr "Input personalized goal as percentage or points (e.g., 50% or 150)"
+
+#: exercise/templates/exercise/save_personalized_points_goal.html
+msgid "REQUIRED_POINTS"
+msgstr "Points required to pass for module"
```
Comment (ihalaij1) [inline locale/en/LC_MESSAGES/django.po:1274]: Where is this message used? I don't see it other files than the translations.

File: locale/fi/LC_MESSAGES/django.po [hunk 7]
```diff
@@ -747,6 +747,41 @@ msgstr "myöhästymissakko"
...
+msgstr "Aseta pistetavoite moduulille"
+
+#: exercise/templates/exercise/personalized_points_goal_modal.html
+msgid "LABEL_POINTS_GOAL_INPUT"
+msgstr "Anna pistetavoite prosentteina tai pisteinä (esim. 50% tai 150)"
+
+#: exercise/templates/exercise/personalized_points_goal_modal.html
+#: exercise/templates/exercise/_user_results.html
+msgid "REQUIRED_POINTS"
+msgstr "Vaaditut pisteet"
```
Comment (ihalaij1) [inline locale/fi/LC_MESSAGES/django.po:769]: Where is this message used? I don't see it other files than the translations.

File: locale/fi/LC_MESSAGES/django.po [hunk 8]
```diff
@@ -747,6 +747,41 @@ msgstr "myöhästymissakko"
...
+
+#: exercise/templates/exercise/personalized_points_goal_modal.html
+msgid "PERSONALIZED_POINTS_MODAL_FAILURE"
+msgstr "Pistetavoitteen asettaminen epäonnistui"
+
+msgid "PERSONALIZED_POINTS_GOAL"
+msgstr "Pistetavoite"
+
+msgid "PERSONALIZED_POINTS_MODAL_LESS_THAN_REQUIRED"
+msgstr "Pistetavoite ei voi olla vähemmän kuin läpipääsyyn vaaditut pisteet"
```
Comment (ihalaij1) [inline locale/fi/LC_MESSAGES/django.po:783]: Can you change `läpipääsyyn` to `läpäisyyn`?

File: locale/fi/LC_MESSAGES/django.po [hunk 9]
```diff
@@ -4091,7 +4126,7 @@ msgstr "Pisteet"
 
 #: exercise/templates/exercise/_points_progress.html
 msgid "POINTS_TO_PASS"
-msgstr "Läpäisyyn vaadittavat pisteet"
+msgstr "Vaaditut pisteet"
```
Comment (ihalaij1) [inline locale/fi/LC_MESSAGES/django.po:4129]: Why was this changed? I like the previous message because there are other messages like `läpäisyyn vaadittavat tehtävät`, and `läpäisyyn vaadittavat moduulit`, which it matches with.

**Review by ihalaij1 — CHANGES_REQUESTED** [review]
Nice!

A couple of comments from me, and a question:
Is it currently possible to remove points goals at all?
If not, it should be added as an option for the user.
Maybe setting the goal to zero should remove it completely?

File: exercise/templates/exercise/personalized_points_goal_modal.html [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+			{% translate "PERSONALIZED_POINTS_MODAL_SUCCESS" %}
+		</div>
+		<div class="alert alert-warning" id="warning-alert" style="display: none;">
+			{% translate "PERSONALIZED_POINTS_MODAL_FAILURE" %}
+		</div>
+		<div class="alert alert-danger" id="danger-alert" style="display: none;">
+			{% translate "PERSONALIZED_POINTS_MODAL_LESS_THAN_REQUIRED" %}
+		</div>
+		<h4 class="title" id="pointsGoalModalLabel">{% translate "PERSONALIZED_POINTS_MODAL_TITLE" %} {{ module.name|parse_localization}}</h4>
+	</div>
```
Comment (ihalaij1) [inline exercise/templates/exercise/personalized_points_goal_modal.html:22]: Can we have quotes around the module name here?

File: locale/fi/LC_MESSAGES/django.po (hunk 9 above)
Comment (mikaelGusse) [reply locale/fi/LC_MESSAGES/django.po:4129]: The original message flowed over the tooltip which looked particularly nasty now that there are three things on it with the points goal. So in production it is currently like this 
image.
Do you think this is ok as a shorter version or is the longer version preferrable even with the overflowing?

Comment (ihalaij1) [reply locale/fi/LC_MESSAGES/django.po:4129]: Okay, I see the problem. You can keep the change.

File: locale/fi/LC_MESSAGES/django.po [hunk 11]
```diff
@@ -747,6 +747,46 @@ msgstr "myöhästymissakko"
...
+
+#: exercise/templates/exercise/personalized_points_goal_modal.html
+msgid "PERSONALIZED_POINTS_MODAL_NOT_NUMBER"
+msgstr "Pistetavoite tulee antaa joko prosentteina tai pisteinä (esim. 50% or 150)"
+
+msgid "PERSONALIZED_POINTS_GOAL"
+msgstr "Pistetavoite"
+
+msgid "PERSONALIZED_POINTS_MODAL_LESS_THAN_REQUIRED"
+msgstr "Pistetavoite ei voi olla vähemmän kuin moduulin läpipääsyyn vaaditut pisteet"
```
Comment (ihalaij1) [inline locale/fi/LC_MESSAGES/django.po:782]: Could you change `läpipääsyyn` --> `läpäisyyn` so that the wording matches other similar messages.

File: exercise/static/exercise/personalized_points_goal.js [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+        const inputValue = $inputField.val().trim();
+
+        const isNumber = !isNaN(inputValue) && inputValue !== '';
+        const isPercentage = inputValue.endsWith('%') && !isNaN(inputValue.slice(0, -1));
+
+        if (!isNumber && !isPercentage) {
+            $('#validation-alert').show();
+            setTimeout(function() {
+                $('#validation-errors-alert').hide();
+            }, 5000);
```
Comment (ihalaij1) [inline exercise/static/exercise/personalized_points_goal.js:18]: This shows a message `#validation-alert` and hides `#validation-errors-alert`.
Meaning, `#validation-alert` never gets hidden.

**Review by ihalaij1 — CHANGES_REQUESTED** [review]
Good job on the fixes!

Is it possible to check if there exists a points goal when clicking the button, and if it exists, show the Remove-button? If there is no points goal the button wouldn't be shown.

Also, if a previous goal exists, it would be nice if it was auto-filled into the points goal input field.

File: exercise/templatetags/exercise.py [hunk 13]
```diff
@@ -64,6 +64,34 @@ def _get_toc(context, student=None):
...
+
+
+@register.simple_tag
+def get_points_goal(module: int, student: int) -> int:
+    student = UserProfile.objects.get(id=student)
+    module = CourseModule.objects.get(id=module)
+    try:
+        return StudentModuleGoal.objects.get(module=module, student=student).personalized_points_goal_points
+    except StudentModuleGoal.DoesNotExist:
+        return False
```
Comment (ihalaij1) [inline exercise/templatetags/exercise.py:92]: Isn't the return value type of this function `Union[int, bool]`?

**Review by ihalaij1 — CHANGES_REQUESTED** [review]
Can the `Remove` button be hidden after it has been clicked? Currently, it's possible to click it more than once, which results in a failure message.

**Review by murhum1 — CHANGES_REQUESTED** [review]
Added some comments and change requests; might still be missing some, but will check the rest tomorrow!

File: exercise/templates/exercise/_points_progress.html [hunk 14]
```diff
@@ -1,16 +1,24 @@
...
 	{% translate "RESULTS_OF_SOME_ASSIGNMENTS_ARE_CURRENTLY_HIDDEN" %}
 </p>
 {% endif %}
 <div class="progress" data-toggle="tooltip" data-html="true" data-placement="bottom"
-    title="{% translate 'POINTS' %}: &lt;span class='text-nowrap'&gt;{{ formatted_points }} / {{ max }}&lt;/span&gt;{% if required %} {% translate 'POINTS_TO_PASS' %}: &lt;span class='text-nowrap'&gt;{{ required }}&lt;/span&gt;{% endif %}">
-    <div class="progress-bar progress-bar-striped progress-bar-{% if full_score %}success{% elif passed %}warning{% else %}danger{% endif %}"
+    title="{% translate 'POINTS' %}: &lt;span class='text-nowrap'&gt;{{ formatted_points }} / {{ max }}&lt;/span&gt;
+        {% if required %} &lt;br&gt; {% translate 'POINTS_TO_PASS' %}: &lt;span class='text-nowrap'&gt;{{ required }}&lt;/span&gt; {% endif %}
+        {% if personalized_points_module_goal %} &lt;span class='personalized-points-full-text text-nowrap'&gt; &lt;br&gt;{% translate 'PERSONALIZED_POINTS_GOAL' %}: &lt;span class='personalized-points-text text-nowrap'&gt;{{ personalized_points_module_goal_points|floatformat:"0" }}&lt;/span&gt;&lt;/span&gt;{% endif %}"
+>
```
[the remark this answers is not in the document]
Comment (mikaelGusse) [reply exercise/templates/exercise/_points_progress.html:13]: This is indeed much better. Added

File: course/models.py
[the remark this answers is not in the document]
Comment (mikaelGusse) [reply course/models.py:1]: This was there to make it so that the percentage would not have to be calculated elsewhere. Changed it to just have the points in the model for clarity.

**Comment by murhum1** [comment]
I added some more fixes:

- Renamed the instances of `personalized_points_goal_*` to `module_goal_*` to make things more understandable (e.g. `personalized_points_module_goal_points` is really difficult to mentally parse). I still kept references to "personalized points goal" in UI text as they were since it works better in that context.
- Fixed some progress-bar CSS classes to use aplus-progress-bar instead
    - Related to that, fixed e2e test CSS selector
- Squashed migrations into a single one by deleting the migration files and re-running makemigrations

Is the renaming operation ok for you @mikaelGusse? If so, I'll proceed with the merge.

**Review by murhum1 — APPROVED** [review]

**Comment by mikaelGusse** [comment]
Yes. All of these changes seem sensible to me. Go ahead and merge 👍
